x-pack/legacy/plugins/code/server/utils/repository_reference_helper.ts

```diff
@@ -0,0 +1,171 @@
+
+  /**
+   * Deletes the reference from the current namespace to the given repository.
+   *
+   * @param uri The uri of the repository.
+   * @returns True if there is no more references to the repository after the deletion.
+   */
+  deleteReference(uri: string): Promise<boolean>;
+
+  /**
+   * Finds all repository uris of which the current namespace has references.
+   *
+   * @returns uris the current namespace references to.
+   */
+  findReferences(): Promise<string[]>;
+}
+
+/**
+ * A factory function helps to create a appropriate helper instance.
+ *
+ * @param client A saved objects client instance.
+ * @returns An helper instance.
+ */
+export function getReferenceHelper(client: SavedObjectsClientContract): RepositoryReferenceHelper {
+  return new SharedObjectReferenceHelper(client);
+}
+
+class SharedObjectReferenceHelper implements RepositoryReferenceHelper {
+  constructor(private readonly client: SavedObjectsClientContract) {}
+
+  /**
+   * TODO better way to access the current space
+   */
+  private namespace(): string {
+    // @ts-ignore
+    return this.client.spaceId;
+  }
+
+  async createReference(uri: string): Promise<boolean> {
+    let obj: SavedObject<any> | undefined;
+    try {
+      obj = await this.client.get(SAVED_OBJ_REPO_REF, uri);
+    } catch (e) {
+      obj = undefined;
+    }
+    if (obj === undefined) {
+      await this.client.create(
+        SAVED_OBJ_REPO_REF,
+        {
+          uri,
+          namespaces: [this.namespace()],
+        },
+        {
+          id: uri,
+        }
+      );
+      return true;
+    } else {
+      const namespaces = new Set((obj.attributes as RepositoryReferences).namespaces);
+      if (namespaces.has(this.namespace())) {
+        throw new Error(`Reference to [${uri}] from [${this.namespace()}] already exists`);
+      }
+      namespaces.add(this.namespace());
+      await this.client.update(
+        SAVED_OBJ_REPO_REF,
+        uri,
+        {
+          uri,
+          namespaces: Array.from(namespaces),
+        },
+        {
+          version: obj.version,
+        }
+      );
+      return false;
+    }
+  }
+
+  async deleteReference(uri: string): Promise<boolean> {
+    const obj = await this.client.get(SAVED_OBJ_REPO_REF, uri);
+    const refs = obj.attributes as RepositoryReferences;
+    const namespaces = new Set(refs.namespaces);
+    namespaces.delete(this.namespace());
+    if (namespaces.size === 0) {
+      // TODO delete should also support versioning
+      await this.client.delete(SAVED_OBJ_REPO_REF, uri);
+      return true;
+    } else {
+      await this.client.update(
+        SAVED_OBJ_REPO_REF,
+        uri,
+        {
+          uri,
+          namespaces: Array.from(namespaces),
+        },
+        {
+          version: obj.version,
+        }
+      );
+      return false;
+    }
+  }
+
+  async ensureReference(uri: string): Promise<void> {
+    if (!(await this.hasReference(uri))) {
+      throw new Error();
```

Comment:
nit: can we provide a reason for the error?

Reply:
Fixed